ADD: wine-mono url & download

mmtrt opened the pull request:
This adds wine mono url checks and downloads it to wine cache to be installed which is required by some .net frameworks.

lucyllewy commented:
I don't think this is needed. WINE should automatically fetch this when required.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,21 @@ init_wine() {
       mv "$WINEPREFIX" "${WINEPREFIX}.bak-$(date +%s)"
   fi
 
+  if [[ -n "${MONO_URL:-}" ]]; then
```
I would prefer sommelier-core provides this URL if possible. So users can just set a flag "I want mono to be installed".

mmtrt replied:
we could add it this would require to have each mono version url and check it with flag envar having mono version on it still will require updating in case of future versions.

lucyllewy replied:
it should be a build time application of the sommelier core snaps and be irrelevant to end consumer snaps.

merlijn-sebrechts replied:
The required mono version depends on the Wine version. We don't know at build time what Wine version we'll be using since this depends on the content snap connected at runtime. Therefore, I don't think we can add `wine-mono` at build time.

Ideally, we automatically check at runtime what Mono version is required and download the right version. So we have a flag `SOMMELIER_INSTALL_MONO=1`. If this flag is set, we check which version of Mono is required by Wine and download that version.

However, I'm currently not finding a good way to extract the required mono version, though. Our options:

* A table [in this page](https://wiki.winehq.org/Mono) maps wine version to mono version, but I'd rather not parse HTML.
* It's in [appwiz.dll](https://github.com/wine-mirror/wine/blob/e909986e6ea5ecd49b2b847f321ad89b2ae4f6f1/dlls/appwiz.cpl/addons.c#L60) but I don't know of a reliable way to extract that version from the dll itself.
* Getting the wine version is easy to do from the CLI so we could add a table mapping Wine version to Mono version in sommelier-core. But then we need to update this table each time we release a new platform snap.

mmtrt replied:
* Checking mono version from file `appwiz.dll.so` in `wine 3.x` yeilds the result but on newer version doesn't have `appwiz.dll.so` due to PE conversion but on non PE build same file doesn't have string. So its not usefull for us following this method.
*  we could add all mono versions in script then match mono versions with wine and assign.
* we can check wine version from file `libwine.so.1.0` from content snaps without using `wine --version`

merlijn-sebrechts replied:
> * we could add all mono versions in script then match mono versions with wine and assign.
> * we can check wine version from file libwine.so.1.0 from content snaps without using wine --version

Ok, then I think this is the best solution currently. You can implement this, unless @diddledani has objections. Just curious, what is the advantage of using `libwine.so.1.0` instead of `wine --version`?

mmtrt replied:
we grep version string from lib file on the otherhand we are exec wine inorder to check version.

merlijn-sebrechts reviewed (commented):
Like Dani says, can you provide more info on why this is needed? If I understand it correctly, in a normal installation, either Wine downloads Mono, or Mono is provided by the OS. Using one of those methods would be preferred over implementing our own download logic.

* Normally, Wine asks the user whether to download and install mono if it's not found:
  
  ![image](https://user-images.githubusercontent.com/1492981/133938392-ea86efc5-7bef-4c59-afab-f55542d3f8bb.png)

  This is not a good UX since the user can click "cancel" and shoot themselves in the foot. But since Wine already implements the logic to download and install Mono, is there a way we can just use Wine's logic without giving the user the option to "cancel"?
* Can Wine use Mono from the Ubuntu installation? In that case, can we include mono in the Wine platform snap? This page seems to suggest we can just bundle mono in the platform snap: https://wiki.winehq.org/Mono This would be preferred since it reduces the "first start" time.

mmtrt commented:
I think this approach is better but could be even better with what I suggested in changes comment.

Shipping mono in platform snaps as snaps are already big enough users complain.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
+
+            4.21|5.0.3)
+              MONO_VER="4.9.4"
+              ;;&
```
What does this `&` do?

mmtrt replied:
From `man bash` Using `;;&` in place of `;;` causes the shell to test the next pattern list in the statement, if any, and execute  any associated  list  on a successful match, continuing the case statement execution as if the pattern list had not matched.

merlijn-sebrechts replied:
Aha, thanks! I didn't know that.

So this introduces a bug because of the default condition `*)`. If wine is version `4.21`, it will set `MONO_VER` to `4.9.4`, but keep checking the next cases, eventually ending up on the case `*)`, which changes `MONO_VER` to `6.3.0` and shows the error.

You can just remove the `&` and your code will work correctly. `4.21|5.0.3)` works correctly without `&` because it is a _single condition_ from the point of view of the `case` statement. Even though it checks for two values.

merlijn-sebrechts replied:
As an example: the following outputs `4.9.4`

```shell
WINE_VER="5.0.3"

case "$WINE_VER" in
4.21|5.0.3)
  MONO_VER="4.9.4"
  ;;
*)
  MONO_VER="6.3.0"
  echo "Error can't find wine version! setting wine-mono version to $MONO_VER"
  ;;
esac
```

While the following outputs `6.3.0`:

```shell
WINE_VER="5.0.3"

case "$WINE_VER" in
4.21|5.0.3)
  MONO_VER="4.9.4"
  ;;&
*)
  MONO_VER="6.3.0"
  echo "Error can't find wine version! setting wine-mono version to $MONO_VER"
  ;;
esac
```

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
+
+            5.22|6.0.1)
+              MONO_VER="5.1.1"
+              ;;&
```
What does this `&` do?

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
+              MONO_VER="5.1.1"
+              ;;&
+
+            6.14|6.15|6.16|6.17)
```
Also add a case for `*` so we have some fallback if wine version is not known. Ideally, this prints a warning and assigns the highest-known mono version (`6.3.0` in this case)

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
+    if [[ $(grep -c mscoree "$WINEPREFIX"/user.reg 2&>/dev/null) -eq 0 ]]; then
+        WINE_VER=$(strings "$WINE_PLATFORM"/lib/libwine.so.1 | grep "wine-[\.0-9]" | cut -d "-" -f 2)
+        # For future reference MONO_VER https://source.winehq.org/source/dlls/appwiz.cpl/addons.c
+          case "$WINE_VER" in
```
indentation is wrong for the entire case block. Should be two spaces less.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
 
+  if [[ -n "${SOMMELIER_INSTALL_MONO:-}" ]]; then
+    if [[ $(grep -c mscoree "$WINEPREFIX"/user.reg 2&>/dev/null) -eq 0 ]]; then
+        WINE_VER=$(strings "$WINE_PLATFORM"/lib/libwine.so.1 | grep "wine-[\.0-9]" | cut -d "-" -f 2)
```
Indentation is wrong for most of this block; should be two spaces less. Indentation is 2 spaces, not 4.

merlijn-sebrechts commented on scripts/sommelier:
```diff
@@ -67,6 +67,49 @@ init_wine() {
+      if [[ ! -f "${XDG_CACHE_HOME}/wine/${MONO_MSI}" ]]; then
+        wget "${MONO_URL}" -O "${XDG_CACHE_HOME}/wine/${MONO_MSI}" 2>&1 | \
+        perl -p -e '$| = 1; s/^.* +([0-9]+%) +([0-9,.]+[GMKB]) +([0-9hms,.]+).*$/\1\n# Downloading... \2 (\3)/' | \
+        yad --progress --title="${MONO_MSI}" --width=400 --center --no-buttons --auto-close --auto-kill --on-top --no-escape
```
Change title to `--title="Downloading ${MONO_MSI}"`

merlijn-sebrechts reviewed (changes requested):
Looks good! I left some final comments. I haven't tested it though, I'll do that this weekend.

mmtrt commented:
I think we need to check for `winedlloverrides` having `mscoree` otherwise would be useless to download it if have override set by consumer snap.

merlijn-sebrechts commented:
> I think we need to check for winedlloverrides having mscoree otherwise would be useless to download it if have override set by consumer snap.

It would be useful to throw a warning when the dll override and SOMMELIER_INSTALL_MONO are both set, but I don't think it's critical. We should definitely add an explanation about this to the docs, though.

mmtrt commented on scripts/sommelier:
```diff
@@ -67,6 +67,53 @@ init_wine() {
   fi
 
+  if [[ -n "${SOMMELIER_INSTALL_MONO:-}" ]]; then
+    if [[ $(grep -c mscoree "$WINEPREFIX"/user.reg 2&>/dev/null) -eq 0 ]]; then
```
if wineprefix has dotnet20 installed this still equals 0 and then re downloads mono we have to place more checks for this.

mmtrt commented on scripts/sommelier:
```diff
@@ -67,6 +67,53 @@ init_wine() {
+        MONO_VER="6.4.0"
+        ;;
+      esac
+      MONO_URL="https://dl.winehq.org/wine/wine-mono/${MONO_VER}/wine-mono-${MONO_VER}.msi"
```
mono versions starting from 5.x.x have filename change with addition of `-x86`

`wine-mono-5.x.x-x86.msi`

mmtrt committed:
ADD: wine-mono url & download

This adds wine mono url checks and downloads it to wine cache to be installed which is required by some .net frameworks.

jnsgruk commented:
Closing for now as likely stale - happy to review/support if any further appetite to complete this.
